lib/web_ui/lib/src/engine/text/layout_service.dart

```diff
@@ -270,6 +287,145 @@ class TextLayoutService {
     }
   }
 
+  ui.TextDirection get _paragraphDirection =>
+      paragraph.paragraphStyle.effectiveTextDirection;
+
+  /// Positions the boxes in the given [line] and takes into account their
+  /// directions, the paragraph's direction, and alignment justification.
+  void _positionLineBoxes(EngineLineMetrics line, {
+    required bool withJustification,
+  }) {
+    final List<RangeBox> boxes = line.boxes;
+    final double justifyPerSpaceBox = withJustification
+      ? _calculateJustifyPerSpaceBox(line)
+      : 0.0;
+
+    int i = 0;
+    double cumulativeWidth = 0.0;
+    while (i < boxes.length) {
+      final RangeBox box = boxes[i];
+      if (box.boxDirection == _paragraphDirection) {
+        // The box is in the same direction as the paragraph.
+        box.startOffset = cumulativeWidth;
+        box.lineWidth = line.width;
+        if (box is SpanBox && box.isSpaceOnly) {
+          box._width += justifyPerSpaceBox;
```

Comment:
Do we need to skip increasing the width of the last space-only box? Or maybe the last space-only box should have zero width if the text is justified? :thinking:

Reply:
Good intuition! I was going by what the browser does. But now I checked Flutter, and you're right, it collapses the width of trailing spaces!

Reply:
FWIW, Flutter always collapses trailing spaces (regardless of which alignment is used).

Reply:
Actually, I (partially) lied.

During text painting/rendering, Flutter treats trailing spaces as if they were zero-width (e.g. no background).

During text editing (cursor placement & selection highlighting), Flutter treats trailing spaces the same as other spaces.

| Left | Right | Justify |
|-|-|-|
| ![image](https://user-images.githubusercontent.com/1278212/159339726-d63a4784-62e4-428c-8585-be8a4a5a9d6f.png) | ![image](https://user-images.githubusercontent.com/1278212/159339806-bf0bc4a5-3597-4893-a550-0804dbafd0a3.png) | ![image](https://user-images.githubusercontent.com/1278212/159339901-fa41705e-678f-4251-9b91-20a54decc2fa.png) |

I believe what's happening during text editing is they have no special case for trailing spaces, but they clamp the results of `getBoxesForRange` to the width of the paragraph.

Reply:
Done (minus the clamping part).

Reply:
Filed an issue to implement the clamping: https://github.com/flutter/flutter/issues/100567